subprojects/tooling-api-builders/src/test/groovy/org/gradle/tooling/internal/provider/runner/ClientProvidedPhasedActionRunnerTest.groovy

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal.provider.runner
+
+import org.gradle.BuildListener
+import org.gradle.BuildResult
+import org.gradle.StartParameter
+import org.gradle.api.BuildCancelledException
+import org.gradle.api.internal.GradleInternal
+import org.gradle.initialization.BuildEventConsumer
+import org.gradle.internal.invocation.BuildController
+import org.gradle.internal.service.ServiceRegistry
+import org.gradle.tooling.internal.protocol.InternalBuildActionFailureException
+import org.gradle.tooling.internal.protocol.InternalBuildActionVersion2
+import org.gradle.tooling.internal.protocol.InternalBuildCancelledException
+import org.gradle.tooling.internal.protocol.InternalPhasedAction
+import org.gradle.tooling.internal.provider.BuildActionResult
+import org.gradle.tooling.internal.provider.BuildClientSubscriptions
+import org.gradle.tooling.internal.provider.ClientProvidedPhasedAction
+import org.gradle.tooling.internal.provider.PhasedBuildActionResult
+import org.gradle.tooling.internal.provider.serialization.PayloadSerializer
+import org.gradle.tooling.internal.provider.serialization.SerializedPayload
+import spock.lang.Specification
+
+class ClientProvidedPhasedActionRunnerTest extends Specification {
```

Comment:
❌ This comment applies to all unit tests: Can you please reduce the number of Mocks to only those interactions you are trying to verify? Everything else should be Stubs (or just the real implementation if it is a value object). 

The reason behind this is that mocks lock you down to a very specific implementation, so that refactorings can become extremely painful.